driver/src/deposit_driver.rs

```diff
@@ -86,9 +87,167 @@ pub fn run_deposit_listener<D, C>(db: &D, contract: &C) -> Result<(), DriverErro
         }
     } else {
         println!("All deposits are already processed");
     }
-    Ok(())
+    Ok(false)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::models::Hashable;
+    use crate::contract::tests::SnappContractMock;
+    use crate::models::tests::test_flux;
+    use crate::db_interface::tests::DbInterfaceMock;
+    use mock_it::Matcher::*;
+
+    #[test]
+    fn applies_current_state_if_unapplied_and_enough_blocks_passed() {
+        let slot = U256::from(1);
+        let state_hash = H256::zero();
+        let deposits = vec![test_flux(1,1), test_flux(1,2)];
+        let state = models::State {
+            stateHash: format!("{:x}", state_hash),
+            stateIndex: 1,
+            balances: vec![100; (models::TOKENS * 2) as usize],
+        };
+
+        let contract = SnappContractMock::new();
+        contract.get_current_deposit_slot.given(()).will_return(Ok(slot));
+        contract.has_deposit_slot_been_applied.given(slot).will_return(Ok(false));
+        contract.has_deposit_slot_been_applied.given(slot - 1).will_return(Ok(true));
+        contract.creation_block_for_deposit_slot.given(slot).will_return(Ok(U256::from(10)));
+        contract.get_current_block_number.given(()).will_return(Ok(U256::from(34)));
+        contract.deposit_hash_for_slot.given(slot).will_return(Ok(deposits.hash()));
+        contract.get_current_state_root.given(()).will_return(Ok(state_hash));
+        contract.apply_deposits.given((slot, Any, Any, Any)).will_return(Ok(()));
+
+        let db = DbInterfaceMock::new();
+        db.get_deposits_of_slot.given(1).will_return(Ok(deposits));
+        db.get_current_balances.given(state_hash).will_return(Ok(state));
+
+        assert_eq!(run_deposit_listener(&db, &contract), Ok(true));
+    }
+
+    #[test]
+    fn does_not_apply_if_highest_slot_already_applied() {
+        let slot = U256::from(1);
+        let state_hash = H256::zero();
+
+        let state = models::State {
+            stateHash: format!("{:x}", state_hash),
+            stateIndex: 1,
+            balances: vec![100; (models::TOKENS * 2) as usize],
+        };
+
+        let contract = SnappContractMock::new();
+        contract.get_current_state_root.given(()).will_return(Ok(state_hash));        
+        contract.get_current_deposit_slot.given(()).will_return(Ok(slot));
+        contract.has_deposit_slot_been_applied.given(slot).will_return(Ok(true));
+
+        contract.get_current_block_number.given(()).will_return(Ok(U256::from(11)));
+        contract.creation_block_for_deposit_slot.given(slot + 1).will_return(Ok(U256::from(10)));
+        contract.deposit_hash_for_slot.given(slot + 1).will_return(Ok(H256::zero()));
+
+        let db = DbInterfaceMock::new();
+        db.get_current_balances.given(state_hash).will_return(Ok(state));
+
+        assert_eq!(run_deposit_listener(&db, &contract), Ok(false));
+    }
+
+    #[test]
+    fn does_not_apply_if_highest_slot_too_close_to_current_block() {
```

Comment:
This seems alright, but have we kept in mind the fact that this deposit slot becomes "apply-able" in the event that is has already been filled to its capacity?

Reply:
Yeah, this is a good observation. At the moment the driver always waits for the blocks to pass. I'll create an issue. This is something we should probably solve in a separate fix PR, that has a unit test and the fix.